### PR TITLE
Implements some 32-bit instructions (VBIC, VTST, VSRA)

### LINKED ISSUE #1005: Unimplemented VBIC ARM32 instruction
Requires the ARM32 VBIC instruction (immediate variant).

```
The instruction at 0x0000000003a30f94 (opcode 0xf2c0617f) is undefined!
```

#### Description:

```
VBIC.I32	Q11, #0xF
```

## Required by:
* https://github.com/Ryujinx/Ryujinx-Games-List/issues/149
* https://github.com/Ryujinx/Ryujinx-Games-List/issues/313
* https://github.com/Ryujinx/Ryujinx-Games-List/issues/314
* https://github.com/Ryujinx/Ryujinx-Games-List/issues/392

### LINKED ISSUE #1345: Unimplemented VBIC.I16 instruction
Requires the VBIC.I16 instruction.  Note: this occurs even with https://github.com/Ryujinx/Ryujinx/pull/1192 implemented.

```
The instruction at 0x000000000176f6d4 (opcode 0xf3876b7f) is undefined!
```

**Description:**
```
vbic.i16 q3, #0xff00
```

## Required by: 
Obtained the above opcode error after testing with https://github.com/Ryujinx/Ryujinx/pull/1192, which got the below game past prior missing opcode crashes.
https://github.com/Ryujinx/Ryujinx-Games-List/issues/1499

### PR DESCRIPTION
**Pull request by valx76**
Fixes #1005
Fixes #1345 

Ryujinx/Ryujinx-Games-List#313
Ryujinx/Ryujinx-Games-List#314
Ryujinx/Ryujinx-Games-List#392
Ryujinx/Ryujinx-Games-List#750
Ryujinx/Ryujinx-Games-List#1523
Ryujinx/Ryujinx-Games-List#1579
Star Wars Episode I Racer
=> Those games now run

Ryujinx/Ryujinx-Games-List#145
Ryujinx/Ryujinx-Games-List#149
Ryujinx/Ryujinx-Games-List#1852
=> Those games don't start yet (still have missing instructions)

Ryujinx/Ryujinx-Games-List#304
Ryujinx/Ryujinx-Games-List#511
Ryujinx/Ryujinx-Games-List#626
Ryujinx/Ryujinx-Games-List#1362 / Ryujinx/Ryujinx-Games-List#182
Ryujinx/Ryujinx-Games-List#1634
=> I don't own those games but they are reported below as not working

---
Thanks to @gdkchan for his great help!
Thanks to @EmulationFanatic for his great testing/reporting efforts!

### CODE REVIEW

File: ARMeilleure/Instructions/InstEmitSimdHelper32.cs [hunk 1]
```diff
@@ -349,6 +370,16 @@ public static void EmitVectorBinaryOpZx32(ArmEmitterContext context, Func2I emit
         {
             EmitVectorBinaryOpI32(context, emit, false);
         }
+        
+        public static void EmitVectorImmTernaryOpZx32(ArmEmitterContext context, Func3I emit)
```
Comment (gdkchan) [inline ARMeilleure/Instructions/InstEmitSimdHelper32.cs:374]: Those should be located above the new `EmitVectorImmTernaryOpI32` function you added.

File: ARMeilleure/Instructions/InstEmitSimdLogical32.cs [hunk 2]
```diff
@@ -114,6 +162,19 @@ public static void Vorr_II(ArmEmitterContext context)
 
             context.Copy(GetVecA32(op.Qd), res);
         }
+        
+        public static void Vtst(ArmEmitterContext context)
+        {
+            EmitVectorBinaryOpZx32(context, (op1, op2) =>
+            {
+                if (context.BitwiseAnd(op1, op2).Value == 0L)
```
Comment (gdkchan) [inline ARMeilleure/Instructions/InstEmitSimdLogical32.cs:170]: This is not valid at all. The bitwise AND operation will be only evaluated when the code is actually executed. `Value` in this case will always be zero. The only cases where the `Value` is not 0 is when the operand is a constant, which is not the case here since the AND result is not constant.

File: ARMeilleure/Instructions/InstEmitSimdShift32.cs [hunk 3]
```diff
@@ -128,6 +128,30 @@ public static void Vshrn(ArmEmitterContext context)
...
+                    {
+                        return Const(op3.Type, 0);
+                    }
+
+                    return context.Add(op1, context.ShiftRightUI(op3, Const(shift)));
+                });
+            }
+            else
+            {
+                EmitVectorImmTernaryOpSx32(context, (op1, op2, op3) => context.Add(op1, context.ShiftRightSI(op3, Const(Math.Min(maxShift, shift)))));
```
Comment (gdkchan) [inline ARMeilleure/Instructions/InstEmitSimdShift32.cs:152]: Something seems off here, `op2` is not used at all.

Comment (valx76) [reply ARMeilleure/Instructions/InstEmitSimdShift32.cs]: Changed the method to use a binary operation and used the naming we talked about

**Comment by GoffyDude** [comment]
**Opcode 0xf2c0617f Family**
Tested the initial targeted games here, which all now boot. though they all have the same graphic issues; they seem to share a common graphics engine.
[STRIKERS1945](https://github.com/Ryujinx/Ryujinx-Games-List/issues/313) - now boots and goes in-game. has missing graphics.
[STRIKERS1945II](https://github.com/Ryujinx/Ryujinx-Games-List/issues/314) - now boots and goes in-game. has missing graphics
[SamuraiAces for Nintendo Switch](https://github.com/Ryujinx/Ryujinx-Games-List/issues/392) - now boots and goes in-game. has missing graphics
[GUNBIRD for Nintendo Switch](https://github.com/Ryujinx/Ryujinx-Games-List/issues/750) - now boots and goes in-game. has missing graphics
[SolDivide for Nintendo Switch](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1523) - now boots and goes in-game. has missing graphics


**Other missing opcode 32-bit games tested:**
I also tested some other 32-bit games that have opcode errors, but none of them were resolved with this PR.

[Pocket Rumble](https://github.com/Ryujinx/Ryujinx-Games-List/issues/958) still crashes - opcode 0xe128f00c
[Goat Simulator](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1295) - still crashes - opcode 0xeea00ba1
[Cabela's: The Hunt - Championship Edition](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1305) - still crashes - opcode 0xeea20a41
[Calculation Castle: Greco's Ghostly Challenge "Addition"](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1330) - still crashes - 0xf2d443a1
[Sky Gamblers: Storm Raiders](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1499) - still crashes - opcode 0xf3876b7f

**Comment by GoffyDude** [comment]
Monster Hunter Generations Ultimate - still crashes - opcode 0xe1011042

**Comment by SeraUX** [comment]
[Moero Chronicle Hyper](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1362) - still crashes - opcode 0xf2d223a1

**Comment by EmulationFanatic** [comment]
https://github.com/Ryujinx/Ryujinx-Games-List/issues/1579 is fixed with this PR

**Comment by riperiperi** [comment]
Are you going to add unit tests for these? I can go through the code while looking at the reference manual, but it would be nice to compare against Unicorn too.

**Comment by valx76** [comment]
Hey @riperiperi, if needed I can check how it has been done currently in Ryujinx and add some unit tests.
I just thought the unit tests were required for the 64-bit instructions but not the 32-bit ones..

**Comment by EmulationFanatic** [comment]
[DEADLY PREMONITION Origins](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1634) - still crashes - opcode 0xf3f90760

**Comment by gdkchan** [comment]
Needs rebase + version increment due to PTC.

**Comment by EmulationFanatic** [comment]
[This War of Mine: Complete Edition](https://github.com/Ryujinx/Ryujinx-Games-List/issues/626) - still crashes - opcode 0xf3f0a5e8

**Comment by EmulationFanatic** [comment]
Star Wars Episode I Racer, which had the missing opcode 0xf25201f0 error, is fixed with this PR.

**Comment by EmulationFanatic** [comment]
[Duke Nukem 3D: 20th Anniversary Edition](https://github.com/Ryujinx/Ryujinx-Games-List/issues/1852) - still crashes - opcode 0xf2d223a1

**Comment by valx76** [comment]
@gdkchan I've rebase the PR and updated the PTC.
I still need to add some unit tests to compare with Unicorn as @riperiperi mentionned above.

**Comment by gdkchan** [comment]
Added tests for the newly implemented instructions, and fixed errors in the implementation, including:
- Fix VTST ignoring `size`. Due to that, it only did 64-bit tests and produced incorrect results.
- Fix VBIC SSE2 path. x86 PANDN instruction inverts the first operand, not the second. Swapping the operands is required to match Arm instruction.
- Fix VSRA. 0 was being returned when shift was equal to the max value (8, 16, 32 or 64), ignoring the addition step.

Hope you don't mind.
All new tests now pass with SSE2 = on and off.

**Comment by valx76** [comment]
@gdkchan I don't mind, thank you so much for your help, once again!
Checked and understood the edits you've made, so all good! :)

**Comment by valx76** [comment]
This update is fixing the issue with the VBIC.I16 instruction (issue #1345)

**Comment by LDj3SNuD** [comment]
```
Ryujinx.Tests\Cpu\CpuTestSimdLogical32.cs:
  Improved Vbic_Vbif_Vbit_Vbsl_Vand_Vorr_Veor_I Test;
  Improved Vbic_Vorr_II Test, Added I16 data type variants;
  Improved Vtst Test.

Ryujinx.Tests\Cpu\CpuTestSimdShImm32.cs:
  Improved Vsra_Vrshr_Vshr_Imm Test.

ARMeilleure\Decoders\OpCodeTable.cs:
  Fixed Vorr_II opcode table entry that caused the added test to fail.

Bumped the PPTC version.

Nits.
```

Hope you don't mind!

lgtm, thanks.

**Comment by valx76** [comment]
Thanks @LDj3SNuD ! I had a discussion with @gdkchan about the VORR immediate implementation, for which I did the fix as well but forgot to push.. My bad lol. Thank you! :)

**Comment by LDj3SNuD** [comment]
Fixed conflict in `InstEmitSimdHelper32.cs` and bumped PPTC version.
